src/transformers/utils/bitsandbytes.py

```diff
@@ -264,29 +264,32 @@ def get_keys_to_not_convert(model):
     else:
         tied_keys = sum(tied_params, [])
     has_tied_params = len(tied_keys) > 0
-
-    # Check if it is a base model
-    is_base_model = not hasattr(model, model.base_model_prefix)
-
-    # Ignore this for base models (BertModel, GPT2Model, etc.)
-    if (not has_tied_params) and is_base_model:
+    if not has_tied_params:
+        # if the model has no tied parameters, we check if the output embeddings are the same as the input embeddings
+        # if they are the same, it just means that the model is using the same weights for the input and output
+        if id(model.get_output_embeddings()) == id(model.get_input_embeddings()):
+            has_tied_params = True
+    # In addition to tied-parameters, we will also keep the "lm-head" in full precision for numerical stability reasons
+    output_embeddings = model.get_output_embeddings()
+    # If no output-embeddings are found, and no tied parameters, we have to quantize the whole model
+    if output_embeddings is None and not has_tied_params:
+        logger.warning(
+            "No output embeddings and tied parameters found, will quantize the whole model. "
+            "This might lead to numerical instability if word/output embeddings are quantized. "
+            "If this is not intended, please check tied parameters definition in your model architecture "
+            "or override `get_output_embeddings()` to return correct output embeddings in your model class."
+        )
         return []
 
-    # otherwise they have an attached head
-    list_modules = list(model.named_parameters())
-    list_last_module = [list_modules[-1][0]]
-
-    # add last module together with tied weights
-    intersection = set(list_last_module) - set(tied_keys)
-    list_untouched = list(set(tied_keys)) + list(intersection)
-
-    # remove ".weight" from the keys
-    names_to_remove = [".weight", ".bias"]
-    filtered_module_names = []
-    for name in list_untouched:
-        for name_to_remove in names_to_remove:
-            if name_to_remove in name:
-                name = name.replace(name_to_remove, "")
-        filtered_module_names.append(name)
-
+    # tie_keys contains ".weight" and ".bias" suffixes
+    list_untouched = [name.replace(".weight", "").replace(".bias", "") for name in tied_keys]
+    # The output_embeddings is usually (but not always) the last module in the model
+    for name, module in model.named_modules():
+        # It's ok to check IDs here even if output_embeddings is None
+        if id(module) == id(output_embeddings):
+            list_untouched.append(name)
+            break  # there should be only one output_embeddings module
```

Comment:
I think it is possible to have multiple tied weights (e.g. T5) but not sure - would it be possible to quickly double check the test you have designed with a t5 model such as `google/flan-t5-xl` ? 🙏

Reply:
yes, there are multiple tied weights.  I have added to the test( Roberta-large). But this line only checks the output-layer(i.e., LM_head). And there should be only one output_embedding?  Not sure the design principle behind the function `get_output_embeddings()`